bbot/core/helpers/web/envelopes.py

```diff
@@ -0,0 +1,352 @@
+
+    def set_subparam(self, key=None, value=None, recursive=True):
+        envelope = self
+        if recursive:
+            envelope = self.final_envelope
+
+        # if there's only one value to set, we can just set it directly
+        if envelope.singleton:
+            envelope.subparams["__default__"] = value
+            return
+
+        # if key isn't specified, use the selected subparam
+        if key is None:
+            key = self.selected_subparam
+        if key is None:
+            raise ValueError(f"{self} -> {envelope}: No subparam selected")
+
+        data = envelope.unpacked_data(recursive=False)
+        for segment in key[:-1]:
+            data = data[segment]
+        data[key[-1]] = value
+
+    @property
+    def name(self):
+        return self.__class__.__name__
+
+    @property
+    def num_envelopes(self):
+        num_envelopes = 0 if self.transparent else 1
+        if self.end_format:
+            return num_envelopes
+        for envelope in self.subparams.values():
+            with suppress(AttributeError):
+                num_envelopes += envelope.num_envelopes
+        return num_envelopes
+
+    @property
+    def summary(self):
+        if self.transparent:
+            return ""
+        self_string = f"{self.friendly_name}"
+        with suppress(AttributeError):
+            child_envelope = self.unpacked_data(recursive=False)
+            child_summary = child_envelope.summary
+            if child_summary:
+                self_string += f" -> {child_summary}"
+
+        if self.selected_subparam:
+            self_string += f" [{'.'.join(self.selected_subparam)}]"
+        return self_string
+
+    def to_dict(self):
+        return self.summary
+
+    def __str__(self):
+        return self.summary
+
+    __repr__ = __str__
+
+
+class HexEnvelope(BaseEnvelope):
+    """
+    Hexadecimal encoding
+    """
+
+    friendly_name = "Hexadecimal-Encoded"
+
+    ignore_exceptions = (ValueError, UnicodeDecodeError)
+
+    def _pack(self, s):
+        return s.encode().hex()
+
+    def _unpack(self, s):
+        return bytes.fromhex(s).decode()
+
+
+class B64Envelope(BaseEnvelope):
+    """
+    Base64 encoding
+    """
+
+    friendly_name = "Base64-Encoded"
+
+    ignore_exceptions = (binascii.Error, UnicodeDecodeError, ValueError)
+
+    def unpack(self, s):
+        # it's easy to have a small value that accidentally decodes to base64
+        if len(s) < 8 and not s.endswith("="):
+            raise ValueError("Data is too small to be sure")
+        return super().unpack(s)
+
+    def _pack(self, s):
+        return base64.b64encode(s.encode()).decode()
+
+    def _unpack(self, s):
+        return base64.b64decode(s).decode()
+
+
+class URLEnvelope(BaseEnvelope):
+    """
+    URL encoding
+    """
+
+    friendly_name = "URL-Encoded"
+
+    def unpack(self, s):
+        unpacked = super().unpack(s)
+        if unpacked == s:
+            raise Exception("Data is not URL-encoded")
```

Comment:
This should probably be ValueError

Reply:
fixed in 8f8a2223578b6a77146ff311f9f49f004ef491ad